docs/proposals/suggestion.md

```diff
@@ -0,0 +1,462 @@
+# Suggestion CRD Design Document
+
+Table of Contents
+=================
+
+   * [Suggestion CRD Design Document](#suggestion-crd-design-document)
+   * [Table of Contents](#table-of-contents)
+      * [Background](#background)
+      * [Goals](#goals)
+      * [Non-Goals](#non-goals)
+      * [Design](#design)
+         * [Kubernetes API](#kubernetes-api)
+         * [GRPC API](#grpc-api)
+         * [Workflow](#workflow)
+            * [Example](#example)
+      * [Algorithm Supports](#algorithm-supports)
+         * [Random](#random)
+         * [Grid](#grid)
+         * [Bayes Optimization](#bayes-optimization)
+         * [HyperBand](#hyperband)
+         * [BOHB](#bohb)
+         * [TPE](#tpe)
+         * [Anneal](#anneal)
+         * [SMAC](#smac)
+
+Created by [gh-md-toc](https://github.com/ekalinin/github-markdown-toc)
+
+## Background
+
+Katib makes suggestions long-running in v1alpha2 and v1alpha3. And the suggestions need to communicate with katib manager to get experiments and trials from katib-db. This design hurts high availability. 
+
+Thus we proposed a new design to implement a CRD for suggestion and remove katib-db from main workflow. The new design simplifies the implmentation of experiment and trial controller, and makes katib Kubernetes native.
+
+This document is to illustrate the details of the new design.
+
+## Goals
+
+- Propose the Suggestion CRD.
+- Propose new GRPC API for Suggestion service.
+- Suggest the approaches to implement suggestion algorithms.
+
+## Non-Goals
+
+- Metrics collection (See [Metrics Collector Design Document](./metrics-collector.md))
+- Database-related refactor
+
+## Design
+
+### Kubernetes API
+
+```go
+type SuggestionSpec struct {
+	//Name of the algorithm
+	AlgorithmName string `json:"algorithm_name"`
+
+	// Number of suggestions requested
+	Suggestions int `json:"suggestions"`
+
+	//Algorithm settings set by the user in the experiment config
+	AlgorithmSettings []AlgorithmSetting `json:"algorithm_settings,omitempty"`
+}
+
+type AlgorithmSetting struct {
+	Name  string `json:"name,omitempty"`
+	Value string `json:"value,omitempty"`
+}
+
+type SuggestionStatus struct {
+	// Suggestion results
+  Assignments []TrialAssignment `json:"assignments,omitempty"`
+  
+  // Algorithm settings set by the algorithm.
+	AlgorithmSettings []AlgorithmSetting `json:"algorithm_settings,omitempty"`
+
+	Conditions []SuggestionCondition `json:"conditions,omitempty"`
+	// include all common fields
+
+}
+
+type TrialAssignment struct {
+	// Suggestion results
+	Assignments []ParameterAssignment `json:"assignments,omitempty"`
+	Name        string                `json:"name,omitempty"`
```

Comment:
Just for clarity. Are we planning to this Name field internally or in the future?

Reply:
Name is optional, I think. When the suggestion updates results, it does not contain name. When the experiment controller creates the trial for the assignment, it will assign name for the `TrialAssignment`

I think we do need a field to know if the assignment is evaluated by a trial.